docs/architecture/adr-059-evidence-composition-and-lifecycle.md

```diff
@@ -0,0 +1,304 @@
+
+  ValidateBasic() error
+  String() string
+}
+```
+
+```go
+type DuplicateVoteEvidence struct {
+  VoteA *Vote
+  VoteB *Vote
+
+  timestamp time.Time // taken from the block time
+}
+```
+
+Tendermint has now introduced a new type of evidence to protect light clients from being attacked. This `LightClientAttackEvidence` is vastly different to `DuplicateVoteEvidence` in that it is physically a much different size containing a complete signed header and validator set. It is formed within the light client, not the consensus engine and requires a lot more information from state to verify (`VerifyLightClientAttack(commonHeader, trustedHeader *SignedHeader, commonVals *ValidatorSet)`  vs `VerifyDuplicateVote(chainID string, pubKey PubKey)`). Finally it batches validators together as opposed to having individual evidence. This evidence stretches the existing mould that was used to accomodate new types of evidence and has thus caused us to reconsider how evidence should be formatted and processed.
+
+```go
+type LightClientAttackEvidence struct { // proposed struct in spec
+  ConflictingBlock *LightBlock
+  CommonHeight int64
+  Type  AttackType     // enum: {Lunatic|Equivocation|Amnesia}
+
+  timestamp time.Time // taken from the block time at the common height
+}
+
+```
+
+## Possible Approaches for Evidence Composition
+
+### Individual framework
+
+Evidence remains on a per validator basis. This causes the least disruption to the current processes but requires that we break `LightClientAttackEvidence` into several pieces of evidence for each malicious validator. This not only has performance consequences in that there are n times as many database operations and that the gossiping of evidence will require more bandwidth then necessary (by requiring a header for each piece) but it potentially impacts our ability to validate it. In batch form, the full node can run the same process the light client did to see that 1/3 validating power was present in both the common block and the conflicting block whereas this becomes more difficult to verify individually without opening the possibility that malicious validators forge evidence against innocent . Not only that, but `LightClientAttackEvidence` also deals with amnesia attacks which unfortunately have the characteristic where we know the set of validators involved but not the subset that were actually malicious (more to be said about this later). And finally splitting the evidence into individual pieces makes it difficult
+
+#### An example of a possible implementation path
+
+We would ignore amnesia evidence (as individually it's hard to make) and revert to the initial split we had before where `DuplicateVoteEvidence` is also used for light client equivocation attacks and thus we only need `LunaticEvidence`. We would also most likely need to remove `Verify` from the interface as this isn't really something that can be used.
```

Comment:
I'm confused about why Verify isn't useful anymore - can you say more? Is it because you can no longer verify every evidence type?

Reply:
It's just because the data required (i.e. the arguments needed) to have a generalised verify function is very broad. It's still possible if we wanted to but we would have to pass both the block store and state store as arguments which kind of raises the barriers of verifying say DuplicateVoteEvidence when for this particular evidence we just need a valset and a chainID

Reply:
Ah, got it, hmm. It _feels_ like `Verify` is something that we should be able to call on all evidences. I wonder if there's another way to access the block store and state store without passing them in as parameters.

Reply:
In the spec it is something like `func IsValid(lcaEvidence LightClientAttackEvidence, bc Blockchain) boolean` in which they pass the entire blockchain

Reply:
lol. _here, have a blockchain_